core/primitives/src/utils.rs

```diff
@@ -427,12 +424,9 @@ pub fn to_timestamp(time: DateTime<chrono::Utc>) -> u64 {
 pub fn get_num_seats_per_shard(num_shards: NumShards, num_seats: NumSeats) -> Vec<NumSeats> {
     (0..num_shards)
         .map(|i| {
-            let remainder = num_seats % num_shards;
-            let num = if i < remainder as u64 {
-                num_seats / num_shards + 1
-            } else {
-                num_seats / num_shards
-            };
+            let remainder = num_seats.wrapping_rem(num_shards);
```

Comment:
Use of a `wrapping` doesn’t actually help anything here. `NumShards` and `NumSeats` are both unsigned, and so the only failure mode to look out for here is caused by `rhs == 0`.

`wrapping_rem` and `wrapping_div` still panic when `rhs` is `0`, it seems, so this change doesn't actually achieve anything and might be better off allowing the lint with a comment.

I would much rather prefer `checked_rem`/`checked_div` and propagate the `Option` to the caller though.

Reply:
in this case `rhs` can't be `0`, the function wouldn't panic on any input, hence using `Option` can be misleading here

I've reverted this change and added ` #[allow(clippy::integer_arithmetic)`

Reply:
> in this case `rhs` can't be `0`, the function wouldn't panic on any input

Sorry if I’m being obtuse here, but what would prevent me or anybody else from calling this function and pass `0` for `num_shards`? The `NumShards` is just a type alias to u64, so there really isn’t anything I can see that would enforce this invariant, other than implicit semantic meaning behind the number of shards.

To be fair, `panic` _is_ an appropriate mechanism for handling these kinds of programming errors (which is why reverting and `#[allow]` of the lint seems fine to me.) But to expand on my thinking, if I was looking for a proof for this kind of claim, I would be expecting a statically enforced approach with something like `std::num::NonZero`; or disallow allow passing in `NumShard` as an argument; or perhaps using a representation where `0` means 1 shard, `1` means 2 shards.

Reply:
passing `num_shards=0` doesn't result in panic right now because it makes `0..num_shards` range empty hence code inside of `.map` is not executed and the function just returns an empty vector.

Reply:
Ah, of course :facepalm: